### ISSUE TITLE
I found the ~ dimensional syntax in reshape difficult to comprehend 

### ISSUE THREAD

**Issue by Inc0n** [marcoheisig/Petalisp#11]
following the example.lisp
To create an array that is 4x4:
```
(compute (reshape 2 (~ 6 9 ~ 6 9)))
```
I realized the pattern after:
```
(compute (reshape 2 (~ 0 9 ~ 0 9)))
```
Which returned a 10x10 array

In contrast, if the ~ dimensional syntax had something akin to:
```
;; for a 4x4 
(compute (reshape 2 (~ 4 4)))
```
I would have immediately caught on.
Using two number to represent each dimension seems a bit redundant.

**Comment by marcoheisig** [comment]
The reason for this is that Petalisp goes a long way in not forcing people to chose a particular array starting index.

If `(~ 4)` would expand into `(~ 0 1 3)`, that would make the starting index of zero more official, and insult all people (especially mathematicians), that think it should expand into `(~ 1 1 4)`.

But you raise a very good point.  Since Lisp arrays are numbered starting from zero anyway, the genie is already out of the bottle.  So how about the following notation:

- `(~ N)` denotes all indices from zero below N.  `(~ 4 ~ 4)` would denote the 4x4 array from your example.  The only question is what should be the space denoted by `(~ -9)`?

- `(~ A B)`and `(A STEP B)` work as usual, such that one-element ranges can be denoted by `(K K)` or `(K 1 K)`.

Of course this would be a breaking change.  But that is precisely why i haven't announced an official 1.0 release yet.

A related issue that then pops up is how to deal with empty shapes.  So far, the syntax did not allow such shapes.  But now, `(~ 0)` won't contain any elements.  Implementation-wise, empty-shapes don't pose a problem.  But once I open this can of worms, I might also introduce some syntax for ranges and shapes with exclusive boundaries, like `(~ A (B))`.  Except that introducing such a notation is not straightforward since `~` is currently implemented as a function.

What do you think?

**Comment by sirherrbatka** [comment]
Well, honestly, trying to please people that can't get over arrays starting with zero feels like an attempt to sell lisp to people that complain about (foo bar) syntax. The issue is that if one has a problem with such a trivial difference how do you expect him to adopt something really different?

Attempt to build an empty array may just signal error.

And I also think that ~ is a one busy symbol.

**Comment by marcoheisig** [comment]
No, @Inc0ns1st3nt is right - the current syntax is not very intuitive.  And so far, Petalisp has improved substantially by incorporating user feedback.

But before I can incorporate this suggestion, I have to get the details right.  Negative indices, exclusive/inclusive boundaries etc.

As for the empty arrays - All built-in operators try to handle empty arrays in an intuitive way.  For example, using `(α OP ...)` returns an empty array when all its arguments are empty arrays, and signals an error when some but not all arguments are empty arrays.  Doing so is a great aid in dealing with corner cases.

**Comment by marcoheisig** [comment]
OK, here is my proposal:

In order to make Petalisp more accessible to users, ranges will work like this:

- `(range)` denotes the empty range - an object I need to introduce.
- `(range K)` denotes the integers from zero to K-1.
- `(range A B)` denotes the integers between A (inclusive) and B (exclusive).
- `(range A B C)` denotes the integers between A (inclusive) and B (exclusive), with a step size of C.
- The sign of the step size is ignored (as usual).  It is an error to supply a step size of zero, except when the resulting range is empty or has only a single element, in which case the step size will be set to one.

The `~` notation is adapted accordingly.  If any of the denoted ranges is the empty range, the result is the empty shape.

I know this change will affect essentially all Petalisp programs in existence, but I hope it is nevertheless a welcome change.  If anyone strongly disagrees with me, please speak up now or remain forever silent :)

**Comment by no-defun-allowed** [comment]
I didn't mind the old range format, but this format could be easier to learn, if one is more familiar with how most programming languages (CL included) handle arrays. They match up with how Common Lisp interprets ranges:

- `(range K)` is the range of indices I can address in a vector of `K` elements.
- `(range A B)` is interpreted in the same way the "range" in `(subseq ... A B)` is interpreted.


I can't think of any precedent for functions that take start, step and end arguments in Common Lisp, but `(range START STEP END)` is somehow nicer to me than `(range START END STEP)`, even though it's a bit harder to parse. Another option is something like Haskell or Miranda's `[start, start+step ... end]` (eg `[1,3..9]` = `[1,3,5,7,9]`), but that might really only make sense with that layer of syntax. Never mind that one then.
CL does usually give options for both inclusive and exclusive ends (eg `(loop for x from 0 to 10 ...)` and `(loop for x from 0 below 10 ...)`, and the type specifiers `(integer 0 10)` and `(integer 0 (10))`), but that seems tricky to include while keeping `range` as a function. `(range 1 :below 10)` or `(range 1 :to 10)` could be implemented with a function, but it's probably another bit harder to parse.

`~` and friends could be complemented with aliases named with words, but I can't think of any words, because they can be mixed in one call, like `(~ 0 3 ~r (range 1 3))`.

**Comment by marcoheisig** [comment]
OK, I just changed the syntax of ranges and shapes like we agreed (e2cb5f5f0b8cb1986186dbd9eff4b9b664d3b61f and 3ec2fa89729a8bd9a97f17d147b9506e8de050a2).

This was quite a lot of work, but I think it is worth it.

Another thing is that @no-defun-allowed gave me the idea of making the shape syntax extensible.  I think I might just do that.  So I'm leaving this issue open until I have resolved that, too.

Any feedback is most welcome, as usual.
